### DISCUSSION TITLE
Create a Blackhole+ with an api that serves the user application (blackhole output) with audioformat etc

### DISCUSSION THREAD

**Discussion by chjan** [ExistentialAudio/BlackHole#755] [Feature Requests]
Hi Devin

Sorry, I need to explain a little before the question/suggestion.

Blackhole is working fine, and as we compare Roon and Audirvana media players, the latter is clearly more audiophile sounding, for this and that reason. Therefore, Audirvana would be the choice. But then, we are also using, and have developed, the SoundPimp product which has the main feature that it in the digital domain will precorrect for acoustic crosstalk (ie in the subsequent analog domain between speakers and ears). Thus, Blackhole on Mac, our own SoundPimp Virtual Audio device on Windows, are mandatory tools to pick up the media stream and route it to SoundPimp

The snag is that, while Roon is kind of "flegmatic" and adapts to whater samplerates etc, Audirvana will not do that, and it forces Blackhole to obey its settings. We see that in AudioMidi Setup.

But, there is no callback API as we are aware of that could be used to force SoundPimp to obey to. Thus, when Audirvana changes the audio parameters, things go awry in SoundPimp as it is unaware of the change. Bummer!

On Windows, this is handles by using the Steinberg ASIO api which will set sample rate and other parameters as they change from track to track. 

As SoundPimp was developed in Java, we needed a better audio machine that the Java Audio Engine, thus we once initiated what became the JASIOHOST, available here on github, and written by Martin Roth with me watching the whole process on the sideline and testing results until it worked. It does, and is today in use in several other appliacation, like the very popular REW, the Room Eq Wisard. BTW, the name Roth is perhaps a funny coincidence, or are you related?

Anyway, here is the suggestion: Would it be possible, or imaginable, to expand Blackhole with a similar (or better; identical) API, i.e. the Steinberg Asio Api? That way, e.g. SoundPimp would connect to Blackhole via its built-in JasioHost-ish and then SoundPimp and other applications would be able to adapt to changed sample rate etc via blackhole callbacks.

Now Audirvana would work in perfection.

Sidenote: JasioHost made Java a fully fledged audio programming language, with down to 64 latency and more. 

I Like the idea more and more, and perhaps it would be possible to call on Martin Roth for a discussion on whether it could be done quite easily.

Thanks for reading!

Kind regards
Carl Henrik Janson

**Comment by chjan** [comment]
my email carl@hdsoundlab.com

**Comment by devinroth** [comment]
Complete coincidence. Unrelated to Martin Roth. Perhaps a long lost cousin.

Sample rate callback are already built into the macOS CoreAudio. And BlackHole supports that. So if you change the sample rate of BlackHole you should be notified via a callback if you register for it.

Devin

**Reply by chjan** [reply]
Thats a challenge on Mac os while using the Java Audio Engine, which is quite an animal from the days of old. SoundPimp also have a JACK api, so perhaps I need to look into that again. It fell apart some years ago when Apple tightened the rope with a sandbox etc. 

Just to add another potential, I imagine Oracle would be very pleased to see a renewed alternative to Java Audio Engine. If the same was done on Linux, according to my suggestion, this solution could be a candidate for the NEW java audio module.

**Comment by devinroth** [comment]
What are you proposing?

Devin

**Reply by chjan** [reply]
It is too much on the fringe of Blackhole, so I send you an email found at your homepage.

**Reply by devinroth** [reply]
I never received your email. Please send again. devinroth@existential.audio
